### PR TITLE
fix: [termdebug] correct K/-/+ mappings check

### PR DESCRIPTION
**Pull request by Shane-XB-Qian**
// and did same like 'saved_mousemodel' chk

### CODE REVIEW

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim [hunk 1]
```diff
@@ -1232,32 +1232,38 @@ def DeleteCommands()
...
     elseif empty(k_map_saved)
-      nunmap K
+      silent! nunmap K
     endif
-    k_map_saved = {}
+    k_map_saved = null_dict
   endif
-  if exists('plus_map_saved')
-    if !empty(plus_map_saved) && !plus_map_saved.buffer
+  if plus_map_saved isnot null_dict
```
Comment (errael) [inline runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1246]:
```suggestion
  if plus_map_saved != null
```
For reference
```
vim9script
var d1: dict<any> = null_dict
var d2: dict<any> = {}
echo 'is d1 null?' d1 == null
echo 'is d2 null?' d2 == null
```
outputs
```
is d1 null? true
is d2 null? false
```

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim [hunk 2]
```diff
@@ -1232,32 +1232,38 @@ def DeleteCommands()
   delcommand Var
   delcommand Winbar
 
-  if exists('k_map_saved')
-    if !empty(k_map_saved) && !k_map_saved.buffer
+  if k_map_saved isnot null_dict
```
Comment (errael) [inline runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1235]:
```suggestion
  if k_map_saved != null
```

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim [hunk 3]
```diff
@@ -1232,32 +1232,38 @@ def DeleteCommands()
...
     elseif empty(plus_map_saved)
-      nunmap +
+      silent! nunmap +
     endif
-    plus_map_saved = {}
+    plus_map_saved = null_dict
   endif
-  if exists('minus_map_saved')
-    if !empty(minus_map_saved) && !minus_map_saved.buffer
+  if minus_map_saved isnot null_dict
```
Comment (errael) [inline runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1257]:
```suggestion
  if minus_map_saved != null
```

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim (hunk 1 above)
Comment (Shane-XB-Qian) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1246]: if you recalled some months ago what we argued in such null/null_obj topic which chris applied your PR,
i am confused now truly null or null_obj which was your taste? :sweat_smile:
anyway, `{} isnot null_dict` also is true, specially it is just keeping it same operation as `saved_mousemodel` did,
so i think ok.

Comment (errael) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1246]: strings are the weirdest of them all when it comes to null handling. With strings you must use `isnot null_string`, as it says at  `:help null-anomalies`
```
Unlike the other containers, an uninitialized string is equal to null.
```

I left out some important cases from the example.
```vim
echo 'is d1 null_dict?' d1 == null_dict
echo 'is d2 nul_dictl?' d2 == null_dict
```
Both return `true`.

`!= null` is easier to understand than `isnot null_dict` (at least for me). It also looks like what you'd expect to see in most any other language; to me, that makes it easier to maintain.

As you say, what's currently there works; just not as maintainable (but that's only my opinion)

**Comment by ubaldot** [comment]
@Shane-XB-Qian Thanks! Could you please add some tests?

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim [hunk 4]
```diff
@@ -1232,32 +1232,38 @@ def DeleteCommands()
...
-    if !empty(k_map_saved) && !k_map_saved.buffer
+  if k_map_saved isnot null_dict
+    if !empty(k_map_saved) && k_map_saved.buffer
+      # pass
+    elseif !empty(k_map_saved) && !k_map_saved.buffer
       nunmap K
       mapset(k_map_saved)
     elseif empty(k_map_saved)
-      nunmap K
+      silent! nunmap K
```
Comment (chrisbra) [inline runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1242]: why the silent!

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim [hunk 5]
```diff
@@ -1232,32 +1232,38 @@ def DeleteCommands()
...
-    if !empty(plus_map_saved) && !plus_map_saved.buffer
+  if plus_map_saved isnot null_dict
+    if !empty(plus_map_saved) && plus_map_saved.buffer
+      # pass
+    elseif !empty(plus_map_saved) && !plus_map_saved.buffer
       nunmap +
       mapset(plus_map_saved)
     elseif empty(plus_map_saved)
-      nunmap +
+      silent! nunmap +
```
Comment (chrisbra) [inline runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1253]: why the silent!

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim (hunk 4 above)
Comment (Shane-XB-Qian) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1242]: there maybe buf local mapping at source, so e.g gdb had no 'K' mapping actually.

Comment (Shane-XB-Qian) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1242]: there maybe buf local mapping at source window, then e.g gdb window had no 'K' mapping actually.

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim (hunk 2 above)
Comment (chrisbra) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1235]: that is not resolved

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim (hunk 4 above)
Comment (chrisbra) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1242]: alright then, thanks. Can you please make the suggested changes from @errael ?

Comment (Shane-XB-Qian) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1242]: that's a confusion to himself, https://github.com/vim/vim/pull/15013#discussion_r1641334643 had explained.
// though you accepted his PR vs/over mine in the past about this null/null_obj topic, but just a code style.
// vs this termdebug you had accepted null_obj for `saved_mousemodel`, a similar flag, so i just kept the same.

Comment (Shane-XB-Qian) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1242]: that's a confusion to himself, https://github.com/vim/vim/pull/15013#discussion_r1641334643 had explained.
// though you accepted his PR vs/over mine in the past about this null/null_obj topic, but just a code style.
// vs this termdebug you had accepted null_obj for `saved_mousemodel`, a similar flag, so i just kept the same.

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim (hunk 2 above)
Comment (Shane-XB-Qian) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1235]: https://github.com/vim/vim/pull/15013#discussion_r1641661003

File: runtime/pack/dist/opt/termdebug/plugin/termdebug.vim (hunk 4 above)
Comment (chrisbra) [reply runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:1242]: ah, okay. So you initialize to null_dict and then testing against null_dict should be correct.

**Pull request closed without merging**